Enhancement of Low Temperature Radiant Models to Allow Multi-Zone Systems

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8050 
 - Another new feature as a result of requests from the Center for the Built Environment (CBE) at the University of California-Berkeley.  Previously, EnergyPlus would require that all surfaces that are part of a single radiant system would be required to be in the same zone and that this zone be the same as the zone being served by the radiant system.  Now, with these changes, the surfaces can be in any zone, do not need to all be in the same zone as each other, and do not need to be in the same zone as the zone being served by the radiant system.  As a result, radiant system surfaces can be controlled by conditions in another zone and all of the surfaces in the radiant system surface group will be controlled identically regardless of the zone in which they exist.  This pull request includes code modifications to allow multi-zone radiant systems, a unit test of a new centralized subroutine that serves all three radiant models, documentation changes, and a new IDF that demonstrates how to do this as well as that this new feature is working in the code.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: NewFeature
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [0] If any diffs are expected, author must demonstrate they are justified using plots and descriptions: no diffs in previously existing IDFs.
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] ~If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix~
 - [x] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [x] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [x] Check any new function arguments for performance impacts
 - [x] Verify IDF naming conventions and styles, memos and notes and defaults
 - [x] If new idf included, locally check the err file and other outputs

RKStrand committed:
Design Document

This commit adds the new feature proposal/design document for this branch.

RKStrand committed:
Merge branch 'develop' into 8050-MultiZoneRadiantSystems

RKStrand committed:
Changes to implement multi-zone radiant systems

This commit includes changes to the code to implement multi-zone control.  The checks on zones have been altered to allow surfaces within a single radiant system to be in different zones.  Initial tests show that surfaces that are all part of the same system are controlling in the same fashion based on temperature output at the source location.

RKStrand committed:
Merge branch 'develop' into 8050-MultiZoneRadiantSystems

RKStrand committed:
Correction, Unit Test, Docs, New IDF

This commit includes a variety of issues.  First, there was a correction to some of the error messages in the new subroutine to make them generic to all systems.  Then, there was the addition of a unit test to verify that the new subroutine is working properly, documentation changes to reflect the ability to simulate radiant systems with surfaces in different zones, and new IDF to show that the new functionality is working as expected.

RKStrand committed:
Merge branch 'develop' into 8050-MultiZoneRadiantSystems

mjwitte commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -1702,7 +1702,7 @@ \subsubsection{Inputs}\label{inputs-8-017}
 \paragraph{Field: Surface \textless{}x\textgreater{} Name}\label{field-surface-x-name-2}
 
-This field is the name of a surface in the zone being conditioned by the radiant system. Base surfaces (e.g., \hyperref[buildingsurfacedetailed]{BuildingSurface:Detailed}), door surfaces and internal mass are valid. Window surfaces are not valid surface types for embedded radiant systems.
+This field is the name of a surface being conditioned by the radiant system. It should be noted that the surfaces listed in this field do \emph{not} need to be present in the zone being served by the radiant system to which this surface is related. This allows radiant systems to span more than one zone and have surfaces controlled by conditions in another zone. When the user does select surfaces  ...
```
The intro paragraph also has a sentence about surface being in the same zone. I'll push up a commit that deletes that.

mjwitte replied:
Actually, I'm going to let you take care of that.

RKStrand replied:
No problem.  It is fixed now and this should show up in the next commit.

mjwitte commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1637,6 +1583,52 @@ namespace LowTempRadiantSystem {
     }
 
+    void RadiantSystemBaseData::errorCheckZonesAndConstructions(bool &errorsFound,
+                                                                std::string const& cAlphaField3,
+                                                                std::string const& alpha3,
```
Nit-picking, but `alpha3` and `alpha1` arguments aren't needed here. They're already available in `this->Name` and `this->ZoneName`.

`cAlphaField3` seems unnecessary as well, it's "Zone Name" and the message would read better as `"...Radiant System in Zone=ZoneABC` anyway.

Finally, maybe it would be useful to add the object type name to the base class struct and set that during input? Or set it as a const in the various specific classes? Might be a good pattern everywhere - for objects to know their class name? @Myoldmopar @mitchute ? Then this wouldn't need any arguments. Looks like [plant components](https://github.com/NREL/EnergyPlus/blob/e884660c04f1cc6f14ff32f105ee650637069b7f/src/EnergyPlus/Plant/Component.hh#L65) have a `TypeOf` variable.

RKStrand replied:
@mjwitte Ok, I've eliminated everything except the object type passed variable and the errorsFound variable.  So, this is much better.  While I agree that it would be nice to keep system type, it would take some work to get that engineered out of the radiant code and to the way you want it.  Undoubtably, it would improve some aspects of the code, and it would be a better pattern.  But is this necessary here and now or after someone looks at all of the code and establishes the "standard"?  Anyway, changes will show up in the next commit.

RKStrand replied:
Actually, let's do this.  I will simply get rid of the need for this variable here.  The error message, instead of being:

"Occurs in " + currentModuleObject + " = " + this->Name

Has been changed to:

"Occurs in Low Temperature Radiant System = " + this->Name

The other error message has also been rewritten, eliminating the need for the system type to be passed.  Since there are only three low temperature radiant system types, it seems reasonable that with the name of the system, the surface, and the zone that a user could easily identify the source of the issue.  So, this should take care of the issue without any need to do a bunch of rewriting/reworking of the code.  Will show up in the next commit.

mjwitte commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1637,6 +1583,52 @@ namespace LowTempRadiantSystem {
+            }
+            
+            // check zone multipliers--these must be the same
+            if (SurfNum == 1) zoneMultipliers = double(DataHeatBalance::Zone(this->ZonePtr).Multiplier * DataHeatBalance::Zone(this->ZonePtr).ListMultiplier);
+            zoneMultipliersSurface = double(DataHeatBalance::Zone(Surface(this->SurfacePtr(SurfNum)).Zone).Multiplier
```
Visual Studio is warning that the integers should be cast before multiplying: `x = double(integer1) * double(integer2)` instead of `x = double(integer1 * integer2)` I know this warning shows up all over the code, but we may as well start clean here.

RKStrand replied:
Fixed.  For both this line and the next two lines of code.  Should show up in the next commit.

mjwitte commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1637,6 +1583,52 @@ namespace LowTempRadiantSystem {
+            zoneMultipliersSurface = double(DataHeatBalance::Zone(Surface(this->SurfacePtr(SurfNum)).Zone).Multiplier
+                                            * DataHeatBalance::Zone(Surface(this->SurfacePtr(SurfNum)).Zone).ListMultiplier);
+            if (std::abs(zoneMultipliers-zoneMultipliersSurface)>zoneMultiplersTolerance) {
+                ShowSevereError("The zone multipliers in " + currentModuleObject + " are not the same for the zones in all of the surfaces");
```
How about "The zone multipliers are not the same for all surfaces contained in this radiant system. . . . .

mjwitte reviewed (commented):
@RKStrand Some initial comments on this. I'll do some testing tomorrow or Monday so there may be other comments.

RKStrand committed:
Merge branch 'develop' into 8050-MultiZoneRadiantSystems

RKStrand committed:
Changes for Review Comments

This commit includes a fix for a merge mistake/typo as well as changes made to address review comments made regarding this pull request.

RKStrand commented:
@mjwitte All of your comments have been addressed as noted above.  A new commit has been made and this branch is at the time of the writing of this comment is up-to-date with develop.

mjwitte commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -577,25 +577,7 @@ namespace LowTempRadiantSystem {
-                    ErrorsFound = true;
-                }
-            }
+            thisRadSys.errorCheckZonesAndConstructions(ErrorsFound);
```
Ah, so crisp and clean, no caffeine.

mjwitte commented on testfiles/RadLoTempHydrHeatCoolMultiZone.idf:
```diff
@@ -0,0 +1,2195 @@
+! RadLoTempHydrHeatCoolMultiZone.idf
```
Here's what is in the error file for this:
```
   ** Warning ** SetUpZoneSizingArrays: Requested Sizing for Zone="EAST ZONE", Zone is not found in the Controlled Zones List
   ** Warning ** SetUpZoneSizingArrays: Requested Sizing for Zone="EAST ZONE", Zone has no thermostat (ref: ZoneControl:Thermostat, et al)
   ** Warning ** SetUpZoneSizingArrays: Requested Sizing for Zone="NORTH ZONE", Zone is not found in the Controlled Zones List
   ** Warning ** SetUpZoneSizingArrays: Requested Sizing for Zone="NORTH ZONE", Zone has no thermostat (ref: ZoneControl:Thermostat, et al)
   ** Warning ** Not all surfaces in same zone for ZoneHVAC:LowTemperatureRadiant:SurfaceGroup = SURFACESINTHREEZONES
   ** Warning ** Not all surfaces in same zone for ZoneHVAC:LowTemperatureRadiant:SurfaceGroup = SURFACESINTHREEZONES
   ** Warning ** A surface referenced in a Low Temperature Radiant System is not in same zone as the radiant system itself
   **   ~~~   ** Surface = ZN002:FLR001
   **   ~~~   ** Surface in Zone=EAST ZONE Radiant System in Zone = WEST ZONE
   **   ~~~   ** Occurs in Low Temperature Radiant System = WEST ZONE RADIANT FLOOR
   **   ~~~   ** If this is intentionally a radiant system with surfaces in more than one thermal zone,
   **   ~~~   ** then ignore this warning message.  Otherwise, check the surfaces in this radiant system.
   ** Warning ** A surface referenced in a Low Temperature Radiant System is not in same zone as the radiant system itself
   **   ~~~   ** Surface = ZN003:FLR001
   **   ~~~   ** Surface in Zone=NORTH ZONE Radiant System in Zone = WEST ZONE
   **   ~~~   ** Occurs in Low Temperature Radiant System = WEST ZONE RADIANT FLOOR
   **   ~~~   ** If this is intentionally a radiant system with surfaces in more than one thermal zone,
   **   ~~~   ** then ignore this warning message.  Otherwise, check the surfaces in this radiant system.
```
1. The Sizing:Zone objects should be deleted for East Zone and North Zone.
2. That's a lot of words for a warning that might be ignorable. 
2a. This one should only fire once per surfacegroup: `** Warning ** Not all surfaces in same zone for ZoneHVAC:LowTemperatureRadiant:SurfaceGroup = SURFACESINTHREEZONES`
2b. Since this is a potentially OK condition, how about wrapping `if (DisplayExtraWarnings)` around the surface-by-surface warnings for not in the same zone.
2c. These lines could move to the surface group warning:
```
   **   ~~~   ** If this is intentionally a radiant system with surfaces in more than one thermal zone,
   **   ~~~   ** then ignore this warning message.  Otherwise, check the surfaces in this radiant system. Use Output:Diagnostics, DisplayExtraWarnings for more details.
```

RKStrand replied:
@mjwitte Looks great, Mike!  Thanks for making that change, for your review, and for getting this merged in!! 😀👍

mjwitte committed:
LowTempRadiantSystem tweak surface warnings

mjwitte reviewed (commented):
Doc changes look good. Experimented with the new idf. Zone multiplier error and internal source error work OK. Unit tests pass.

@RKStrand So CI can run overnight, I went ahead and made the changes discussed below. Let me know if you approve. This is what the err files look like now.
[RadLoTempHydrHeatCoolMultiZone.err.txt](https://github.com/NREL/EnergyPlus/files/5025673/RadLoTempHydrHeatCoolMultiZone.err.txt)
[RadLoTempHydrHeatCoolMultiZone-ExtraWarnings.err.txt](https://github.com/NREL/EnergyPlus/files/5025674/RadLoTempHydrHeatCoolMultiZone-ExtraWarnings.err.txt)

mjwitte reviewed (approved):
@RKStrand CI is happy. I'm going to go ahead and merge this. If you disagree with my changes to the warnings, you can adjust that in your next radiant PR.
